tests/e2e/test_client.go

```diff
@@ -47,97 +46,113 @@ func (q QueryResponse) Array(key string) []QueryResponse {
 	return result
 }
 
-func Querier(t *testing.T, chain *ibctesting.TestChain) func(contract string, query Query) QueryResponse {
-	return func(contract string, query Query) QueryResponse {
+func Querier(t *testing.T, chain *ibctesting.TestChain) func(contract string, query Query) (QueryResponse, error) {
+	return func(contract string, query Query) (QueryResponse, error) {
 		qRsp := make(map[string]any)
 		err := chain.SmartQuery(contract, query, &qRsp)
-		require.NoError(t, err)
-		return qRsp
+		if err != nil {
+			return nil, err
+		}
+		return qRsp, nil
 	}
 }
 
 type TestProviderClient struct {
 	t     *testing.T
-	chain *ibctesting.TestChain
+	Chain *ibctesting.TestChain
 }
 
 func NewProviderClient(t *testing.T, chain *ibctesting.TestChain) *TestProviderClient {
-	return &TestProviderClient{t: t, chain: chain}
+	return &TestProviderClient{t: t, Chain: chain}
 }
 
-func (p TestProviderClient) mustExec(contract sdk.AccAddress, payload string, funds []sdk.Coin) *abci.ExecTxResult {
-	rsp, err := p.Exec(contract, payload, funds...)
-	require.NoError(p.t, err)
-	return rsp
-}
-
-func (p TestProviderClient) Exec(contract sdk.AccAddress, payload string, funds ...sdk.Coin) (*abci.ExecTxResult, error) {
-	rsp, err := p.chain.SendMsgs(&wasmtypes.MsgExecuteContract{
-		Sender:   p.chain.SenderAccount.GetAddress().String(),
+func (p *TestProviderClient) Exec(contract sdk.AccAddress, payload []byte, funds ...sdk.Coin) (*abci.ExecTxResult, error) {
+	rsp, err := p.Chain.SendMsgs(&wasmtypes.MsgExecuteContract{
```

Comment:
can you explain how `SendMsgs` work, i briefly checked and it seems to be some helper func in ibctesting module which I didn't understand. I thought we need to do something like this `wasmd tx wasm instantiate <args>`

Reply:
also if you could explain what is `p.Chain` here, does this start a wasmd node in background?

Reply:
So all of these come from the `ibctesting` [module](https://github.com/CosmWasm/wasmd/tree/main/x/wasm/ibctesting) that allows to write unit tests about IBC protocols. The `p.Chain` is a Cosmos SDK chain defined in the [TestChain](https://github.com/CosmWasm/wasmd/tree/main/x/wasm/ibctesting) struct. You can see it includes a fully fledged `App` and various metadata. `SendMsgs` allows us to send arbitrary message to this chain.

In L70, we send a `MsgExecuteContract` tmessage o the chain. `MsgExecuteContract` is used for submitting a certain message to a certain smart contract. For Babylon SDK, we will be using this command to send a message to BTC staking contract upon each `BeginBlock`, such that BTC staking contract can update the voting power table, tally non-finalised blocks, etc..